plugins/providers/docker/action/network.rb

```diff
@@ -0,0 +1,96 @@
+require 'log4r'
+
+module VagrantPlugins
+  module DockerProvider
+    module Action
+      class Network
+        def initialize(app, env)
+          @app = app
+          @logger = Log4r::Logger.new('vagrant::plugins::docker::network')
+        end
+
+        # TODO: This is not ideal, but not enitrely sure a good way
+        # around this without either having separate config options (which is worse for the user)
+        # or forcing them to care about options for each separate docker network command (which is also
+        # not great for the user). Basically a user wants to say:
+        #
+        # config.vm.network :private_network, <config options>
+        #
+        # We could limit the options to be name spaced simply `docker__option` like
+        #
+        # config.vm.network :private_network, docker__ip: "ip here"
+        #
+        # But then we will need this method to split out each option for creating and connecting networks
+        #
+        # Alternatively we could do
+        #
+        # - `docker__create__option`
+        # - `docker__connect__option`
+        # - ...etc...
+        #
+        # config.vm.network :private_network, docker__connect__ip: "ip here",
+        #                                     docker__create__subnet: "subnet"
+        #
+        # But this will force users to care about what options are for which commands,
+        # but maybe they will have to care about it anyway, and it isn't that big of
+        # a deal for the extra namespacing?
+        #
+        # Extra namespacing puts more effort on the user side, but would allow us
+        # to be 'smarter' about how we set options in code by seeing essnetially
+        # what command the flag/option is intended for, rather than us trying to keep
+        # track of the commands/flags manually in this function.
+        #
+        # @param[Hash] options - options from the network config
+        # @returns[Hash] cli_opts - an array of strings used for the network commnad
+        def parse_cli_arguments(options)
+          cli_opts = {create: [], connect: []}
+
+          # Splits the networking options to generate the proper CLI flags for docker
+          if options[:type] != "dhcp"
+            options.each do |opt, value|
+              vals = opt.to_s.split("__")
+              if vals[0] == "docker" && vals.size == 3
+                cli_opts[vals[1].to_sym].concat(["--#{vals[2]}", value])
+              end
+            end
+          end
```

Comment:
Ok, this is generally the idea I was thinking about in the comments above. Essentially a user would give us:

```ruby
 docker.vm.network :private_network,
        docker__connect__ip: "172.20.128.2", docker__create__subnet: "172.20.0.0/16",
        docker__create__driver: "bridge"
```

And this function would generate the following flags internally to use for the right `docker` commands:

```ruby
{create: ["--subnet", "172.20.0.0/16", "--driver", "bridge"], connect: ["--ip", "172.20.128.2"]}
```

Then later on we can use this hashes keys to invoke the proper `docker network` command with the configured options. Maybe this is a too auto-magical but it seems to work at the moment.

Reply:
The pro to this approach at least is that it decouples Vagrant from having to know every single `docker network` subcommand and all of its flags. The con however is that it shifts the responsibility of the user to know which options have to go with each subcommand, which on the surface might not be super intuitive. However if they're going to use networks in docker, it is likely true that they will have to understand each command and what options it needs to configure a network anyway. Also a pro for this subcommand "namespaced" approach is that it makes sure we don't have any config option collisions between `docker network` subcommand options if we had only gone with the `docker__option` approach.